### ISSUE TITLE
several question about implementation

### ISSUE THREAD

**Issue by w4-magnes** [lucidrains/sinkhorn-transformer#22]
I am currently trying to implement Sparse Sinkhorn attention (non-causal, self-attention, pretrain for MLM task) with tensorflow and I would appreciate it if you could answer several questions about your code.

1. I did not quite understand from the paper what is happening with queries and keys in SortCut, when we cut off most of the blocks.
In your code it seems like you broadcast keys/queries like this 
[number_blocks, block_len, size_per_head]
 -> [:n_sortcut, block_len, size_per_head]
 -> [1, n_sortcut * block_len, size_per_head]
 -> [number_blocks, n_sortcut * block_len, size_per_head]
I understand that pytorch handling expand_dim automatically and memory is preserved, but I do not understand how does the memory consumption stay linear:
the resulting attention scores (dots) are still [number_blocks, block_len, num_sortcut * block_len] which is good but quadratic.
Do I miss something?

2. In the case of SortCut what is the meaning of the softmax(A)*Q?
For example, for full attention outputs are vectors which are weighted average of other vectors;
For local block attention outputs are vectors that are weighted average of other vectors in the block.
How would you interpret the output of SortCut attention layer?

3. In your code you concatenate regular keys and values with permuted/sorted keys and values, but in the paper (3.2. part) it seems like simple addition.
Why is it different from the paper? Actually I can understand concatenation more than summation (because I am confused which attention masks to use in this case: from the regular block of permuted).

4. Seems like you are using concatenated queries and keys as an input to make a SortNet. Maybe I completely missed the point, but I did not find anything about this in the article: the input sequence is used for SortNet instead of its projections.
In fact, for some reason, I cannot make it converge if I use query and keys (maybe for a completely unrelated reasons). Have you tried to compare input sequence vs concat(q,k)?

**Comment by lucidrains** [comment]
@w4-magnes Hi! It is good to hear from you, and glad to know there's someone else looking at this.

1. the memory consumption stays linear with respect to the sequence length if `block_len` is held constant. `(seq_len * n_sortcut * block_len)` The way I did the expanding and concatenation is to fall in line with the non-sort-cut method, and perhaps lead to the confusion. You can simply have the full query sequence length attend to the n_sortcut key/value blocks.

2. The way I interpret the paper is that you are asking each head to rank the importance of contiguous key/values blocks in respect to the queries. So the meaning, in your words, would be that  the outputs are vectors which are weighted average of the most relevant key/value vectors as ranked by the sorting net.

3. The addition in 3.2 is actually addressed by the paper. `The secondary termQi(K)>j)is the standard local attention which is added to the mixture.` Like most other papers have noted, language models still benefit from a local inductive bias. The summation in the equation is on the attention matrix, and not on the key/values pre-attention, as I have done in the code.

4. Nice find! I did deviate from the paper on this, mainly because I was aiming to build a full encoder/decoder architecture from the start. My thought was that only using the input sequence would not be able to address routing of contextual key/values. Let me know if you ever compare the two approaches in your MLM efforts, and I will adjust the framework based on the evidence. I have to admit, I have only tested this on some contrived toy tasks, as most of my work has been on making this work as auto-regressive models.

**Comment by Sharut** [comment]
Hi,
Thanks for the response to @w4-magnes comments. These responses definitely help to understand the code better. 
Please excuse me for this basic query, But I don't seem to understand the dimensions of the vectors well in your code. Can you please add some sort of a documentation explaining what are the input shapes that the code expects or maybe explain it here. Also, according to my understanding, given an input sequence (N_embed, L), we first transform it into {Q,V,D} triplets based on the {W_q(i),W_v(i),W_d(i)}, where i corresponds to the head index. Then we proceed with bucketing each of Q,V and D. So are the dimensions of  q expected to be (head_index, bucket_index, length_bucket, embed_dimension) ??? Some other small queries:
1. What is referred by size_per_head?
2. Why use max_buckets?

**Comment by lucidrains** [comment]
@Sharut Hello! You are correct with your assumption in the first paragraph.

1. can you point me to where this variable is used? (size_per_head)

2. the maximum buckets is needed because the sorting net proposed in the paper is a simple linear layer, with weight matrix dimensions `embed_dimension x N_buckets`, where N_buckets is the maximum number of buckets

**Comment by Sharut** [comment]
Hi, 
Thanks for the prompt response:
1. size_per_head is referred by @w4-magnes. I got confused about this since in the code, you expect query(q) to have a dimension ( bh, t, dim) where h is the number of heads. What is b here? Is it the bucket index??? I don't think this is the case since after bucketing, the output dimension of q = (bh, bucket_index, length_bucket, dim) and there is no bucketing involved initially. 
2. So, basically max number of buckets is the same as n_buckets when the number of vectors (t) is a complete multiple of bucket_size. Does max number of buckets account for variable lengthed buckets in the query or the key?
3. Whats the difference between n_top and n_sortcut?

**Comment by lucidrains** [comment]
@Sharut 

1. Ahh, I see where the confusion is. `b` actually stands for the batch dimension. The bucketing dimension is the second one

2. Yup, max number of buckets is the same as `n_buckets`. Max number of buckets does not account for variable-lengthed buckets for query and key, and neither was this addressed in the paper. This is why I improvised a new solution, to allow for variable number of buckets, as explained https://github.com/lucidrains/sinkhorn-transformer#sinkhorn

**Comment by Sharut** [comment]
Ohhhhhhh...
Thanks a ton! I just confused with the dimension terminologies.
Ahh right. Gotcha!
Once again, Thank you so much for your prompt responses and assistance.

**Comment by lucidrains** [comment]
@Sharut no problem! glad to share what I know and good luck in your studies on this remarkable phenomenon!

**Comment by w4-magnes** [comment]
@lucidrains, thank you for your answers!
I compared the input sequence (as in paper?) and concat(q,k) (as in your code) as an input to SortNet. Mask accuracies are about the same, but for a downstream task using concat(q,k) is a bit better.

1. Is there any particular reason why you did not use concat(q,k,v) as input for SortNet?

2. I see that in your code in the gumbel_sinkhorn function you have `r = log(r)`. 
However, for example in the reference code, I used here: https://github.com/google/gumbel_sinkhorn/blob/master/sinkhorn_ops.py#L141
they do not apply log to the input, instead, they kind of assume that the input is already in log-space (and I cannot find a place where they would apply log).
This makes sense because to apply Sinkhorn iterations we need a matrix with nonnegative values like an output of ReLU, but for log(relu(x)) values could easily be negative.
Do I miss something about this?
